File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.py
class ProductOfNumbers:

    def __init__(self):
        self.st = [1]
        self.cnt = [0]

    def add(self, num: int) -> None:
        if num == 0:
            self.cnt.append(self.cnt[-1] + 1)
            self.st.append(1)
        else:
            self.st.append(num * self.st[-1])
            self.cnt.append(self.cnt[-1])

    def getProduct(self, k: int) -> int:
        if self.cnt[-1] - self.cnt[-(k + 1)] > 0:
            return 0
        else:
            return self.st[-1]//self.st[-(k + 1)]



# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)

File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.java
class ProductOfNumbers {
    private int sz;
    ArrayList<Integer>lt;
    public ProductOfNumbers() {
        lt = new ArrayList<>();
        lt.add(1);
    }
    
    public void add(int num) {
        if(num == 0){
            sz = 0 ;
            lt = new ArrayList<>();
            lt.add(1);
        }else{
            sz ++ ;
            lt.add(lt.getLast() * num);
        }
    }
    
    public int getProduct(int k) {
        if(k > sz) return 0;
        else return lt.get(sz)/lt.get(sz - k);
    }
}

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers obj = new ProductOfNumbers();
 * obj.add(num);
 * int param_2 = obj.getProduct(k);
 */